File: backend-node/src/config/app_factory.js
require('dotenv').config();

const express = require('express');
const cors = require('cors');
const apiRoutes = require('../routes/api_routes');
const { MongoService } = require('../services/mongo_service');

async function createApp(config = {}) {
  const app = express();

  const debug = config.DEBUG || process.env.DEBUG === 'true';
  const mongoUri = config.MONGO_URI || process.env.MONGO_URI || '';

  app.use(cors());
  app.use(express.json());
  app.use(express.urlencoded({ extended: true }));

  if (mongoUri) {
    try {
      const mongo = new MongoService(mongoUri);
      const connected = await mongo.connect();
      if (connected) {
        apiRoutes.setMongoService(mongo);
        console.log('MongoDB connected ✅');
      } else {
        console.warn('MongoDB ping failed — running without DB');
      }
    } catch (exc) {
      console.warn(`MongoDB init failed: ${exc.message}`);
    }
  }

  app.use('/api', apiRoutes);

  app.use((req, res) => {
    res.status(404).json({ error: 'Not found' });
  });

  app.use((err, req, res, next) => {
    console.error('Unhandled error:', err);
    res.status(500).json({ error: 'Internal server error' });
  });

  return app;
}

module.exports = { createApp };

File: backend-python/src/config/app_factory.py
"""
Flask application factory.
"""
import os
from flask import Flask
from flask_cors import CORS

from src.routes import api_bp


def create_app(config: dict = None) -> Flask:
    app = Flask(__name__)
    CORS(app, resources={r"/api/*": {"origins": "*"}})

    # ── Defaults ──────────────────────────────────
    app.config["DEBUG"] = os.getenv("FLASK_DEBUG", "false").lower() == "true"
    app.config["MONGO_URI"] = os.getenv("MONGO_URI", "")

    if config:
        app.config.update(config)

    # ── Optional MongoDB ───────────────────────────
    mongo_uri = app.config.get("MONGO_URI")
    if mongo_uri:
        try:
            from src.services import MongoService
            mongo = MongoService(uri=mongo_uri)
            if mongo.ping():
                app.config["MONGO_SERVICE"] = mongo
                app.logger.info("MongoDB connected ✅")
            else:
                app.logger.warning("MongoDB ping failed — running without DB")
        except Exception as exc:
            app.logger.warning(f"MongoDB init failed: {exc}")

    app.register_blueprint(api_bp)
    return app
